File: cpp/078_coin_partitions.cpp
#include <stdio.h>

int solve_for(int n) {
  printf("Solving for %d\n", n);
  int l[n];
  int i, j;
  for (i = 0; i < n; i++) {
    l[i] = 0;
  }
  for (i = n-1; i >= 0; i--) {
    l[i] = 1;
    for (j = 2*i + 1; j < n; j++) {
      l[j] = (l[j] + l[j - i - 1]) % 1000000;
    }
  }

  for (i = 0; i < n; i++) {
    if (l[i] == 0) {
      return i+1;
    }
  }
  return 0;
}

int main() {
  int m;
  int n = 0;
  for (m = 2; !n; m *= 2) {
    n = solve_for(m);
  }
  printf("%d\n", n);
  return 0;
}

File: python/078_coin_partitions.py
#!/usr/bin/python


def solve_for(n):
  store_mod= False
  print "solving for", n
  l = [0]*n
  for i in xrange(n - 1, -1, -1):
    l[i] = 1
    for j in xrange(2*i + 1, n):
      l[j] += l[j - i - 1]
    if l[n - 1] > 1000000000:
      store_mod = True
    if (store_mod):
      for j in xrange(2*i + 1, n):
        l[j] %= 1000000
  try:
    return l.index(0) + 1
  except ValueError:
    return None

m = 1
n = None
while not n:
  m *= 2
  n = solve_for(m)
print n
